src/engine/engine.cpp

```diff
@@ -53,6 +69,47 @@ Expected<ThreadSafeModule> ExecEngine::optimize_module(
     return std::move(tsm);
 }
 
+void ExecEngine::Optimize(Module &llmod)
+{
+    LoopAnalysisManager LAM;
+    FunctionAnalysisManager FAM;
+    CGSCCAnalysisManager CGAM;
+    ModuleAnalysisManager MAM;
+
+    PassBuilder PB;
+    VectorizerParams::VectorizationFactor = 4;
+
+    PB.registerModuleAnalyses(MAM);
+    PB.registerCGSCCAnalyses(CGAM);
+    PB.registerFunctionAnalyses(FAM);
+    PB.registerLoopAnalyses(LAM);
+    PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
+    FunctionPassManager FPM;
+    FPM.addPass(SCCPPass());
+    FPM.addPass(ADCEPass());
+    FPM.addPass(
+        LoopSimplifyPass());  // needs to be followed by a simplifycfg pass
+    {
+        LoopPassManager LM;
+        LM.addPass(LoopRotatePass());
+        FPM.addPass(createFunctionToLoopPassAdaptor(std::move(LM)));
+    }
+    FPM.addPass(LoopUnrollPass());
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(ADCEPass());
```

Comment:
And do all these in the constructor. I don't think we need to do this every time someone calls the optmize function